cluster-app-values-schema/README.md

```diff
@@ -0,0 +1,205 @@
+# Basic requirements for cluster app values schema
+
+Cluster apps are apps which are used to configure and provision clusters based on Cluster API in the Giant Swarm platform. Like all helm charts, these apps are configured using so-called values YAML files, which are passed to the app as a ConfigMap for cluster creation. The allowed structure of the values YAML is defined by the app's values schema.
+
+This RFC defines basic requirements for all cluster apps provided by Giant Swarm.
+
+## Overview
+
+- [R1: JSON Schema dialect must be specified](#r1)
+- [R2: Schema must explicitly cover all allowed properties](#r2)
+- [R3: Array item schema must be defined](#r3)
+- [R4: Properties must have a title](#r4)
+- [R5: Properties should have descriptions](#r5)
+- [R6: Properties should provide examples](#r6)
+- [R7: Constrain values as much as possible](#r7)
+- [R8: Required properties must be marked as such](#r8)
+
+## Background
+
+TODO: Some more info regarding the rationale.
+
+- In-place documentation of properties
+- Better validation to catch configuration errors before they are applied
+- Generated entry forms
+- Generated information display
+- Enabling processes and communication about explicit contracts instead of implicit behaviour
+
+## Definitions
+
+- **Cluster app**: App (as defined by the Giant Swarm app platform) that allows for provisioning a cluster in a Giant Swarm installation.
+- **JSON Schema**: declarative language that allows to annotate and validate documents that are compatible with JSON.
+- The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://www.rfc-editor.org/rfc/rfc2119.html)
+
+## Requirements
+
+Requirements carry identifiers with prefix `R` and a unique number, for easy referencing.
+
+### R1: JSON Schema dialect must be specified {#r1}
+
+Each cluster app's values schema file MUST specify the schema dialect (also called draft) used via the `$schema` keyword on the top level, with the draft URI as a value.
+
+Example:
+
+```json
+{
+  "$schema": "https://json-schema.org/draft/2020-12/schema",
+  ...
+}
+```
+
+The example above shows an excerpt of a JSON Schema which uses draft 2020-12.
+
+### R2: Schema must explicitly cover all allowed properties {#r2}
+
+All properties that can be used in values MUST be covered in the schema explicitly.
+
+To enforce this and to disable the use of any undefined properties, the keyword `additionalProperties` SHOULD be set to `false` on all object schemas.
+
+The `patternProperties` keyword MUST NOT be used.
```

Comment:
Why do we prohibit the use of `patternProperties`?
If there exists an object that allows arbitrary properties like [here](https://github.com/giantswarm/cluster-cloud-director/blob/main/helm/cluster-cloud-director/values.schema.json#L269), then it would be more specific to use `patternProperties` to constrain how these keys look like. E.g. a regex that enforces all lowercase characters.

Reply:
A valid concern. I guess if we support `additionalProperties`, then we must also support `patternProperties`.

Reply:
Removing the line for now.